DIPs/1NNN-WGB.md

```diff
@@ -0,0 +1,155 @@
+function doesn't escape it.
+
+But if the function is passed more than one reference to the same container,
+one reference can render invalid what the other reference(s) refers to.
+This DIP aims to correct that problem. It's a natural progression after
+DIP25 and DIP1000. It's needed to safely implement Reference Counting.
+
+## Contents
+* [Rationale](#rationale)
+* [Description](#description)
+* [Breaking Changes and Deprecations](#breaking-changes-and-deprecations)
+* [Reference](#reference)
+* [Copyright & License](#copyright--license)
+* [Reviews](#reviews)
+
+## Rationale
+
+Containers cannot be memory safe if they cannot control memory safe
+access to their payloads. Containers cannot be efficient if they cannot
+make available direct references to their payloads. Entreating the user
+to just not do certain things is not reliable and does not scale.
+
+The simplest illustration of the problem:
+
+```
+struct S {
+    byte* ptr;
+    ref byte get() { return *ptr; }
+}
+
+void foo(ref S t, ref byte b) {
+    free(t.ptr);    // frees memory referred to by b
+    b = 4;          // corrupt memory access
+}
+
+void test() {
+    S s;
+    s.ptr = cast(byte*)malloc(1);
+    foo(s, s.get());  // (*)
+}
+```
+The same problem using scope pointers:
+```
+struct S {
+    byte* ptr;
+    byte* get() { return ptr; }
+}
+
+void foo(scope S* t, scope byte* pb) {
+    free(t.ptr);    // frees memory referred to by pb
+    *pb = 4;        // corrupt memory access
+}
+
+void test() {
+    S s;
+    s.ptr = cast(byte*)malloc(1);
+    foo(&s, s.get());  // (*)
+}
+```
+
+D currently has no defense against this problem. It makes checkably memory safe
+reference counting not possible. (Timon Gehr first pointed this out.)
+
+
+## Description
+
+The solution hinges on the recognition that, in the example, two mutable references
+to the same data are passed to the function `foo()`. Generalizing, whenever there are
+multiple references to the same data, and one of them is mutable, the mutable reference
+can be used to invalidate the data the other references refer to, whether they are
+const or not. Therefore, if more than one reference to the same data is passed to
+a function, they must all be const.
+
+This builds on the foundation built and tested by DIP25 and DIP1000 that track lifetimes
+through function calls and their returns, by adding an additional check on data
+already collected by the compiler semantics.
+
+The checks would only be enforced for @safe code.
+
+### Syntax
+
+No syntax changes. It adds additional semantic checks on existing
+constructs.
+
+### Limitations
+
+This is only checking expressions that are function calls. It does not do interstatement
+checking. It does not check non-scope pointers. It is not a complete borrowing/ownership
+scheme, although it is an important step in that direction. For example, the checking
+of scoped pointers can be defeated by using a temporary:
+
+```
+void test() {
+    S s;
```

Comment:
Please add the `struct S` definition to this example.  There are already 2 definitions in this document, and we should be able to see, at a glance, what the definition of `S` is.

Reply:
I don't think the document is long enough for that to be a problem, and repeating the definition again would make the reader suspicious that it is different in some way.

Reply:
There are two, explicitly different definitions of S in that document. Which one is used here is ambiguous, though it can be deduced from context.

(also it uses pointers, even though the document says the DIP is only intended to cover references)